app/controllers/KairosDBProxy.java

```diff
@@ -0,0 +1,349 @@
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+
+/**
+ * A generic proxy controller.
+ *
+ * @author Brandon Arp (barp at groupon dot com)
+ */
+@Singleton
+public class KairosDBProxy extends Controller {
+    /**
+     * Public constructor.
+     *
+     * @param configuration Play configuration to configure the proxy
+     * @param client ws client to use
+     * @param kairosDbClient a KairosDBClient
+     * @param mapper ObjectMapper to use for JSON serialization
+     */
+    @Inject
+    public KairosDBProxy(
+            final Configuration configuration,
+            final WSClient client,
+            final KairosDbClient kairosDbClient,
+            final ObjectMapper mapper) {
+        final String kairosURL = configuration.getString("kairosdb.uri");
+        _client = new ProxyClient(kairosURL, client);
+        _kairosDbClient = kairosDbClient;
+        _mapper = mapper;
+    }
+
+    /**
+     * Proxied status call.
+     *
+     * @return Proxied status response.
+     */
+    public CompletionStage<Result> status() {
+        return proxy();
+    }
+
+    /**
+     * Proxied healthcheck call.
+     *
+     * @return Proxied health check response.
+     */
+    public CompletionStage<Result> healthCheck() {
+        return proxy();
+    }
+
+    /**
+     * Proxied tagNames call.
+     *
+     * @return Proxied tagNames response.
+     */
+    public CompletionStage<Result> tagNames() {
+        return proxy();
+    }
+
+    /**
+     * Proxied tagValues call.
+     *
+     * @return Proxied tagValues response.
+     */
+    public CompletionStage<Result> tagValues() {
+        return proxy();
+    }
+
+    /**
+     * Proxied queryTags call.
+     *
+     * @return Proxied queryTags response.
+     */
+    public CompletionStage<Result> queryTags() {
+        return proxy();
+    }
+
+    /**
+     * Proxied queryMetrics call.
+     *
+     * @return Proxied queryMetrics response.
+     */
+    public CompletionStage<Result> queryMetrics() {
+        return proxy();
+    }
+
+    /**
+     * Proxied version call.
+     *
+     * @return Proxied version response.
+     */
+    public CompletionStage<Result> version() {
+        return proxy();
+    }
+
+    /**
+     * Caching metricNames call.
+     *
+     * @param containing simple string match filter for metric names
+     * @return Cached metric names, filtered by the query string.
+     */
+    public CompletionStage<Result> metricNames(final String containing) {
+        final List<String> metrics = _cache.getIfPresent(METRICS_KEY);
+
+        final CompletionStage<List<String>> response;
+        if (metrics != null) {
+            response = CompletableFuture.completedFuture(metrics);
+        } else {
+            // Refresh the cache
```

Comment:
We should probably background this. Todo?

Reply:
Todo added.